managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2785,6 +2796,12 @@ public void operationComplete(Void result, Stat stat) {
                         log.info("[{}] Deleting offloaded ledger {} from bookkeeper - size: {}", name, ls.getLedgerId(),
                                 ls.getSize());
                         asyncDeleteLedgerFromBookKeeper(ls.getLedgerId());
+                        // in BOOKKEEPER_FIRST read case, we should invalidate the readHandle from bk.
+                        // later, reading this ledger would reacquire readHandle from tiered-storage.
+                        if (config.getLedgerOffloader().getOffloadPolicies().getManagedLedgerOffloadedReadPriority()
+                                == OffloadedReadPriority.BOOKKEEPER_FIRST) {
+                            invalidateReadHandle(ls.getLedgerId());
```

Comment:
There is a race condition. The cursors keep moving forward, and there is a gap between the ledgers check and the ReadHandle invalidation.

For example, one topic's ledger list is [1,2,3,4,5,6], and ledgers [1,2,3,4,5] are offloaded into tiered storage and meet the BookKeeperDeletionLag condition. However, there is a read cursor point to Ledger 3, and the offload read priority is BOOKKEEPER_FIRST.

In the current logic, we will filter out ledger 3 and invalidate the ReadHandle of the ledger list [1,2,4,5]. However, when we do the invalidate action, the cursor already moved from Ledger 3 to Ledger 4 and the cursor will throw NPE when reading data from Ledger 4.

Reply:
make sence. fixed.